### DISCUSSION TITLE
How to provision system dependencies for Python libraries?

### DISCUSSION THREAD

**Discussion by stephan-koenig** [PrairieLearn/PrairieLearn#6130] [Q&A] [answered]
We recently wanted to implement a questions in a course using geoPandas to generate an image. Among its dependencies is a Python library (fiona) that requires GDAL. How can we provision such a system dependency?

- Would you need to use a different Docker image for that course with prairielearn/prairielearn as the base image, and then you install the other dependencies?
- Or is it possible to spawn a non-PL Docker container from within the question that has geoPandas and its dependencies installed to render the image?

**Comment by jonatanschroeder** [comment, marked as answer]
Do you need it to generate a question, in a workspace, or in grading? If you need it to generate a question, you should be able to install it in serverFilesCourse:

```
pip3 install --target serverFilesCourse/ geopandas
```

It may be a good idea to do it inside the PrairieLearn container:

```
docker exec -it prairielearn pip3 install --target /course/serverFilesCourse geopandas
```

If you need it in a workspace or for grading you can certainly create a custom Docker image that has the package installed, but I have a feeling that's not what you need.

**Reply by jonatanschroeder** [reply]
Just to add: for the docker command, replace `prairielearn` with the name of the container running PrairieLearn, and `/course` with the directory where the course in question is mounted (e.g., could be `/course2`, etc.). The purpose of running this inside the PrairieLearn container would be to ensure that the missing dependencies are installed based on what is installed on the PrairieLearn's default packages, rather than what is installed locally on your system.

**Reply by stephan-koenig** [reply]
Great, that worked! Thank you for the explanation!

**Reply by mfbutner** [reply]
@jonatanschroeder I just ran into this same problem so thank you for the answer. I imagine other people will encounter it is as well and so I'd like to add your answer to the official documentation. What section do you think would be best for this?  I'm thinking under [Question Dependencies on the Question Configuration Page](https://prairielearn.readthedocs.io/en/latest/question/#question-dependencies).

**Reply by echuber2** [reply]
It seems like there are two different issues being addressed: the need to install course-specific Python libraries in `serverFilesCourse`, and the way to do this in a local Docker dev container. There are already a few different documentation pages about `serverFilesCourse` as well as about Docker invocations so it's all getting a bit disjointed.

**Reply by jonatanschroeder** [reply]
@mfbutner this was originally discussed in #3892 by @nicknytko, but this was not finished. Maybe you'd like to coordinate with @nicknytko and potentially take over that PR?

About adding it to question dependencies, that section is discussing info.json, so may not be the best place to put it. Maybe a subsection of server.py?
